src/components/calcite-popover/calcite-popover.tsx

```diff
@@ -0,0 +1,217 @@
+
+  @Element() el: HTMLCalcitePopoverElement;
+
+  @State() popper: Popper;
+
+  // --------------------------------------------------------------------------
+  //
+  //  Lifecycle
+  //
+  // --------------------------------------------------------------------------
+
+  componentDidLoad() {
+    this.reposition();
+  }
+
+  componentDidUnload() {
+    this.destroyPopper();
+  }
+
+  // --------------------------------------------------------------------------
+  //
+  //  Public Methods
+  //
+  // --------------------------------------------------------------------------
+
+  @Method() async reposition(): Promise<void> {
+    const {
+      el,
+      placement: componentPlacement,
+      open,
+      popper,
+      referenceElement,
+      xOffset,
+      yOffset
+    } = this;
+
+    const placement =
+      componentPlacement === "vertical" ? "bottom-start" : "auto-start";
+    const offsetEnabled = !!(yOffset || xOffset);
+
+    const modifiers: Popper.Modifiers = {
+      hide: {
+        enabled: false
+      },
+      offset: {
+        enabled: offsetEnabled,
+        offset: `${yOffset}, ${xOffset}`
+      },
+      preventOverflow: {
+        enabled: false
+      }
+    };
+
+    if (popper) {
+      popper.options.placement = placement;
+      popper.options.modifiers = { ...popper.options.modifiers, ...modifiers };
+      popper.scheduleUpdate();
+      return;
+    }
+
+    if (referenceElement && open) {
+      const newPopper = new Popper(referenceElement, el, {
+        eventsEnabled: false,
+        placement,
+        modifiers,
+        onCreate: data => {
+          if (
+            data.originalPlacement === "bottom-start" &&
+            document.body.clientWidth &&
+            data.offsets &&
+            data.offsets.reference &&
+            data.offsets.reference.left > document.body.clientWidth / 2
+          ) {
+            data.instance.options.placement = "bottom-end";
+            data.instance.scheduleUpdate();
+          }
+        }
+      });
+
+      window.addEventListener("resize", newPopper.scheduleUpdate, {
+        passive: true
+      });
+
+      this.popper = newPopper;
+    }
+  }
+
+  @Method() async toggle(): Promise<void> {
+    this.open = !this.open;
+  }
+
+  // --------------------------------------------------------------------------
+  //
+  //  Private Methods
+  //
+  // --------------------------------------------------------------------------
+
+  destroyPopper(): void {
+    const { popper } = this;
+
+    if (popper) {
+      window.removeEventListener("resize", popper.scheduleUpdate);
+      popper.destroy();
+    }
+
+    this.popper = null;
+  }
+
+  // --------------------------------------------------------------------------
+  //
+  //  Render Methods
+  //
+  // --------------------------------------------------------------------------
+
+  render() {
+    return (
+      <Host>
+        <div
+          class={classnames(CSS.container, { [CSS.containerOpen]: this.open })}
```

Comment:
Is there a reason to start wading into CSS-in-JS land here?

Reply:
Nope! will remove. :)

Reply:
This isn't introducing CSS-in-JS - we're still simply placing classes on a JSX tag. This is just a utility for setting classes based on some logic. It helps remove conditional boilerplate for things like "only add this class when this prop is true". [Here's an example of a before an after.](https://www.npmjs.com/package/classnames#usage-with-reactjs)

Reply:
@pr3tori4n there is no boilerplate when you use the api provided by stencil for this:

```
<div class={{
  "open": this.open,
  "container": true
}}>
```

Reply:
@paulcpederson Ah cool, yea Matt mentioned something similar. If that's built-in then we can drop the dependency. There docs for this? I didn't know this was natively available.